adding traffic forecasting example

akhodadadi opened the pull request:
This example shows how to do timeseries forecasting over a graph by combining a graph convolution layer and a LSTM layer. It also shows how to prepare tf.data.Dataset for such applications using `timeseries_dataset_from_array`.

@fchollet could you please review this?

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers, models, optimizers, losses, callbacks, initializers
```
One import per line. Only import modules that you use multiple times; typically only `layers` needs to be imported. Here losses, optimizers, initializers are only used once each so you can do e.g. `keras.optimizers.Adam`, etc.

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+import pandas as pd
+import numpy as np
+import os
+from typing import Tuple, Optional
```
One import per line. Don't import individual symbols, rather import modules (e.g. `import typing`)

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+"""
+To reduce the problem size and make the training faster, we will only
+work with a sample of 26 roads out of the 228 roads in the dataset.
+We have chosen the roads by starting from road 0, choosing the 5 closest
```
What was the motivation for this selection process? It's not obvious, please explain

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+print(f"speeds_array shape={speeds_array.shape}")
+
+"""
+### Data visualization
```
Do you have a live Colab version so we can check out the visualizations?

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+
+def preprocess(data_array: np.ndarray, train_size: float, val_size: float):
+    """Preprocesses data.
+    This function splits data into train/val/test sets and normalizes the data.
```
This line should be the top line. The current top line is not helpful

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+    :return:
+    """
+
+    n_time_steps = data_array.shape[0]
```
Style it: use `num_` as the number prefix, for consistency

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+Next, we create the datasets for our forecasting problem. Let $v^i_t$
+be the speed on road $i$ at time $t$. Then the vector $V_t=[v^1_t,...,v^N_t]$
+will contain the speed for all $N$ roads at time $t$. The forecasting problem
+can be stated as follows: given a sequence $V_{t+1},...,V_{t+T}$ of the
```
No latex notation. Please make sure to review our contributions guidelines: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,655 @@
+
+    dataset = tf.data.Dataset.zip((inputs, targets))
+    if shuffle:
+        dataset = dataset.shuffle(2048)
```
Note that the number here is a number of batches since the data is already batched. 2048 seems excessive.

fchollet reviewed (commented):
Thanks for the PR! This is a very interesting and useful topic, as many datasets feature a combination of graphs and timeseries 👍 

How many lines of code do you end up with? You can use `python tutobooks.py count_loc FILENAME` to check.

akhodadadi commented:
@fchollet Thanks for your valuable comments. I've addressed the issues you raised. I ended up using 297 lines of code. Also, here is a colab version of the notebook:

https://colab.research.google.com/drive/1IjdDrNLOlG0sRabohnu85XaQuNXKrwan#scrollTo=7c3xOjz6ukch

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+    adjacency matrix from the distance matrix. The implementation follows that paper.
+
+    Args:
+        route_distances:
```
Add argument descriptions

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+    Returns:
+        A boolean graph adjacency matrix.
+    """
+    n_routes = route_distances.shape[0]
```
`num_`

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+
+The nodes' representations are obtained by multiplying the features tensor with
+self.weight. Note that
+        self.weight has shape (in_feat, out_feat).
```
Most shapes in the example use lists, prefer using tuples everywhere.

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+        return self.activation(h)
+
+    def call(self, features: tf.Tensor, *args, **kwargs):
+        """
```
Add "Forward pass."

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+
+        return self.activation(h)
+
+    def call(self, features: tf.Tensor, *args, **kwargs):
```
`*args, **kwargs` is not needed

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,656 @@
+
+        self.input_seq_len, self.output_seq_len = input_seq_len, output_seq_len
+
+    def call(self, inputs, *args, **kwargs):
```
`*args, **kwargs` is not needed.

fchollet reviewed (commented):
Thanks for the update! I pushed some copyedits, please pull them first.

akhodadadi commented:
Thanks for the comments. I applied the changes.

fchollet commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,650 @@
+            distance between roads i,j.
+        sigma2: Determines the width of the Gaussian kernel applied to the square distances matrix.
+        epsilon: A threshold specifying if there is an edge between two nodes. Specifically, A[i,j]=1
+            if np.exp(-w2[i,j] / sigma2) >= epsilon and A[i,j]=0 otherwise, where A is the adjacency
```
This applies to multiple places in the docstrings

akhodadadi commented on examples/timeseries/timeseries_traffic_forecasting.py:
```diff
@@ -0,0 +1,650 @@
+    Args:
+        data_array: np.ndarray with shape `(num_time_steps, num_routes)`
+        input_sequence_length: Length of the input sequence (in number of timesteps).
+        forecast_horizon: If multi_horizon=True, the target will be the values of the timeseries for 1 to
```
@fchollet Fixed them

fchollet reviewed (commented):
LGTM, thank you! Please add the generated files.

akhodadadi commented:
@fchollet added the generated files. Thank you for all the feedback.
